### PR TITLE
Expose timeshift api

### PR DESCRIPTION
**Pull request by zigavehovec**
## Problem
The `timeShift` API isn't exposed yet.

## Changes
- Expose `Player.timeShift()`,  `Player.getTimeShift()` and `Player.getMaxTimeShift()`
- Expose timeShift related events: `TimeShiftEvent` and `TimeShiftedEvent`

Note that `TimeShiftEvent.target` has been renamed to `TimeShiftEvent.targetPosition` in the RN world since `target` cannot be mapped to JS objects with `RCTEventEmitter` (both Android and iOS) for some reason.

### CODE REVIEW

File: android/src/main/java/com/bitmovin/player/reactnative/PlayerModule.kt [hunk 1]
```diff
@@ -106,6 +106,18 @@ class PlayerModule(private val context: ReactApplicationContext) : ReactContextB
         }
     }
 
+    /**
+     * Call `.timShift(offset:)` on `nativeId`'s player.
```
Comment (christophaigner) [inline android/src/main/java/com/bitmovin/player/reactnative/PlayerModule.kt:110]:
```suggestion
     * Call `.timeShift(offset:)` on `nativeId`'s player.
```

File: ios/RNPlayerView+PlayerListener.swift [hunk 2]
```diff
@@ -56,7 +56,12 @@ extension RNPlayerView: PlayerListener {
     func onSeeked(_ event: SeekedEvent, player: Player) {
         onSeeked?(event.toJSON())
     }
-
+    func onTimeShift(_ event: TimeShiftEvent, player: Player) {
+        onTimeShift?(event.toJSON())
+    }
+    func onTimeShifted(_ event: TimeShiftedEvent, player: Player) {
+        onTimeShifted?(event.toJSON())
+    }
```
Comment (rolandkakonyi) [inline ios/RNPlayerView+PlayerListener.swift:64]: please add back empty lines

```suggestion

    func onTimeShift(_ event: TimeShiftEvent, player: Player) {
        onTimeShift?(event.toJSON())
    }

    func onTimeShifted(_ event: TimeShiftedEvent, player: Player) {
        onTimeShifted?(event.toJSON())
    }

```

File: ios/PlayerModule.swift [hunk 3]
```diff
@@ -450,4 +462,39 @@ class PlayerModule: NSObject, RCTBridgeModule {
             resolve(self?.players[nativeId]?.isAd)
         }
     }
+    
+    /**
+     The current time shift of the live stream in seconds. This value is always 0 if the active `source` is not a
+     live stream or there is no active playback session.
```
Comment (rolandkakonyi) [inline ios/PlayerModule.swift:468]: We don't have the concept of playback session on iOS.

```suggestion
     The current time shift of the live stream in seconds. This value is always 0 if the active `source` is not a
     live stream or there are no sources loaded.
```

File: src/player.ts [hunk 4]
```diff
@@ -235,6 +235,19 @@ export class Player extends NativeInstance<PlayerConfig> {
...
   };
 
+  /**
+   * Shifts the time to the given `offset` in seconds from the live edge. The resulting offset has to be within the
+   * timeShift window as specified by `maxTimeShift` (which is a negative value) and 0. When the provided `offset` is
+   * positive, it will be interpreted as a UNIX timestamp in seconds and converted to fit into the timeShift window.
+   * When the provided `offset` is negative, but lower than `maxTimeShift`, then it will be clamped to `maxTimeShift`.
+   * Has no effect for VoD.
+   *
+   * Has no effect if there is no active playback session.
```
Comment (rolandkakonyi) [inline src/player.ts:245]: There is no playback session concept on iOS so I would adapt this a bit.

```suggestion
   * Has no effect if the player is unloaded.
```

File: src/player.ts [hunk 5]
```diff
@@ -399,4 +412,20 @@ export class Player extends NativeInstance<PlayerConfig> {
   isAd = (): Promise<boolean> => {
     return PlayerModule.isAd(this.nativeId);
   };
+
+  /**
+   * The current time shift of the live stream in seconds. This value is always 0 if the active `source` is not a
+   * live stream or there is no active playback session.
```
Comment (rolandkakonyi) [inline src/player.ts:418]:
```suggestion
   * live stream or the player is unloaded.
```

File: src/player.ts [hunk 6]
```diff
@@ -399,4 +412,20 @@ export class Player extends NativeInstance<PlayerConfig> {
...
+   */
+  getTimeShift = (): Promise<boolean> => {
+    return PlayerModule.getTimeShift(this.nativeId);
+  };
+
+  /**
+   * The limit in seconds for time shifting. This value is either negative or 0 and it is always 0 if the active
+   * `source` is not a live stream or there is no active playback session.
```
Comment (rolandkakonyi) [inline src/player.ts:426]:
```suggestion
   * `source` is not a live stream or the player is unloaded.
```

**Review by rolandkakonyi — CHANGES_REQUESTED** [review]
I marked some changes with 🚨 as those are functionally broken.

Others are documentation related only.

Note: We missed exposing `LiveConfig` but I think that could be covered in a separate PR later.

File: src/player.ts [hunk 7]
```diff
@@ -399,4 +412,20 @@ export class Player extends NativeInstance<PlayerConfig> {
   isAd = (): Promise<boolean> => {
     return PlayerModule.isAd(this.nativeId);
   };
+
+  /**
+   * The current time shift of the live stream in seconds. This value is always 0 if the active `source` is not a
+   * live stream or there is no active playback session.
+   */
+  getTimeShift = (): Promise<boolean> => {
```
[the remark this answers is not in the document]
Comment (zigavehovec) [reply src/player.ts:420]: I have no idea how I missed this 🙈. Regarding async, should `isAd` then also be async?

File: src/player.ts (hunk 4 above)
Comment (zigavehovec) [reply src/player.ts:245]: What about: 

```suggestion
   * Has no effect if no sources are loaded.
```

File: src/player.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply src/player.ts]: @zigavehovec Yes

File: src/player.ts (hunk 4 above)
Comment (zigavehovec) [reply src/player.ts:245]: I feel like it should be understandable. I'm going with the "Has no effect if no sources are loaded" version.

File: src/player.ts (hunk 5 above)
Comment (zigavehovec) [reply src/player.ts:418]:
```suggestion
   * live stream or no sources are loaded.
```

File: src/player.ts (hunk 6 above)
Comment (zigavehovec) [reply src/player.ts:426]:
```suggestion
   * `source` is not a live stream or no sources are loaded.
```

**Review by rolandkakonyi — APPROVED** [review]

**Pull request merged**